File: kt/Minsk/minsk/src/main/kotlin/dev/phytolizer/minsk/analysis/syntax/SyntaxNode.kt
package dev.phytolizer.minsk.analysis.syntax

import dev.phytolizer.minsk.analysis.TextSpan
import java.io.PrintStream

abstract class SyntaxNode {
    abstract val kind: SyntaxKind
    abstract val children: List<SyntaxNode>

    open val span: TextSpan
        get() {
            val first = children.first().span
            val last = children.last().span
            return TextSpan.fromBounds(first.start, last.end)
        }

    fun prettyPrint() {
        prettyPrint(this, System.out, "", true)
    }

    fun writeTo(writer: PrintStream) {
        prettyPrint(this, writer, "", true)
    }

    private fun prettyPrint(node: SyntaxNode, writer: PrintStream, indent: String, isLast: Boolean) {
        writer.print(indent)
        val marker = if (isLast) {
            "└── "
        } else {
            "├── "
        }
        writer.print(marker)
        if (node is SyntaxToken) {
            writer.print(node)
        } else {
            writer.print(node.kind)
        }
        writer.println()
        val newIndent = indent + if (isLast) {
            "    "
        } else {
            "│   "
        }
        val lastChild = node.children.lastOrNull()
        for (child in node.children) {
            prettyPrint(child, writer, newIndent, child == lastChild)
        }
    }
}

File: cs/Minsk/CodeAnalysis/Syntax/SyntaxNode.cs
using Minsk.CodeAnalysis.Text;

namespace Minsk.CodeAnalysis.Syntax;

public abstract class SyntaxNode
{
    public abstract SyntaxKind Kind { get; }
    public abstract IEnumerable<SyntaxNode> Children { get; }

    public virtual TextSpan Span
    {
        get
        {
            var first = Children.First().Span;
            var last = Children.Last().Span;
            return TextSpan.FromBounds(first.Start, last.End);
        }
    }

    public void PrettyPrint()
    {
        PrettyPrintInternal(this, Console.Out, true, "", true);
    }

    public void WriteTo(TextWriter writer)
    {
        PrettyPrintInternal(this, writer, false, "", true);
    }

    private static void PrettyPrintInternal(SyntaxNode node, TextWriter writer, bool isToConsole, string indent, bool isLast)
    {
        if (isToConsole)
        {
            Console.ForegroundColor = ConsoleColor.DarkGray;
        }

        writer.Write(indent);
        var marker = isLast ? "└───" : "├───";
        writer.Write(marker);

        if (isToConsole)
        {
            Console.ForegroundColor = node is SyntaxToken ? ConsoleColor.Blue : ConsoleColor.Cyan;
        }

        writer.Write(node.Kind);
        if (isToConsole)
        {
            Console.ResetColor();
        }

        if (node is SyntaxToken { Value: { } } t)
        {
            writer.Write($" {t.Value}");
        }

        writer.WriteLine();

        indent += isLast ? "    " : "│   ";
        var children = node.Children.ToArray();
        var lastChild = children.LastOrDefault();
        foreach (var child in children)
        {
            PrettyPrintInternal(child, writer, isToConsole, indent, child == lastChild);
        }
    }
}
